File: Aug/Day23/find-the-minimum-area-to-cover-all-ones-ii.cc
//  Time: O(N^2 & N^2)  ::  Space: O(N * M)

class Solution {
public:
    int minimumArea(vector<vector<int>>& grid, int y1, int y2, int x1, int x2) {
        int n = grid.size(), m = grid[0].size();
        int minX = n, maxX = 0, minY = m, maxY = 0;

        for (int i = y1; i <= y2; i++) {
            for (int j = x1; j <= x2; j++) {
                if (grid[i][j] == 1) {
                    minX = min(minX, i);
                    maxX = max(maxX, i);
                    minY = min(minY, j);
                    maxY = max(maxY, j);
                }
            }
        }
        return (minX <= maxX) ? (maxX - minX + 1) * (maxY - minY + 1) : INT_MAX / 3;
    }

    vector<vector<int>> rotate(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = (n > 0 ? grid[0].size() : 0);
        vector<vector<int>> rotated(m, vector<int>(n, 0));

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                rotated[m - j - 1][i] = grid[i][j];
            }
        }
        return rotated;
    }

    int solve(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = (n > 0 ? grid[0].size() : 0);
        int res = n * m;

        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < m - 1; j++) {
                res = min(res,
                          minimumArea(grid, 0, i, 0, m - 1) +
                          minimumArea(grid, i + 1, n - 1, 0, j) +
                          minimumArea(grid, i + 1, n - 1, j + 1, m - 1));

                res = min(res,
                          minimumArea(grid, 0, i, 0, j) +
                          minimumArea(grid, 0, i, j + 1, m - 1) +
                          minimumArea(grid, i + 1, n - 1, 0, m - 1));
            }
        }

        for (int i = 0; i < n - 2; i++) {
            for (int j = i + 1; j < n - 1; j++) {
                res = min(res,
                          minimumArea(grid, 0, i, 0, m - 1) +
                          minimumArea(grid, i + 1, j, 0, m - 1) +
                          minimumArea(grid, j + 1, n - 1, 0, m - 1));
            }
        }

        return res;
    }

    int minimumSum(vector<vector<int>>& grid) {
        vector<vector<int>> rgrid = rotate(grid);
        return min(solve(grid), solve(rgrid));
    }
};

File: Aug/Day23/find-the-minimum-area-to-cover-all-ones-ii.py
#  Time: O(N^2 & N^2)  ::  Space: O(N * M)
class Solution:
    def minimumArea(self, grid, y1: int, y2: int, x1: int, x2: int
    )  -> int:
        n, m = len(grid), len(grid[0])
        minX, maxX = n, 0
        minY, maxY = m, 0

        for i in range(y1, y2+1):
            for j in range(x1, x2+1):
                if grid[i][j] == 1:
                    minX = min(minX, i)
                    maxX = max(maxX, i)
                    minY = min(minY, j)
                    maxY = max(maxY, j)

        return (maxX - minX + 1) * (maxY - minY + 1) if minX <= maxX else sys.maxsize // 3

    def rotate(self, grid):
        n = len(grid)
        m = len(grid[0]) if n > 0 else 0
        rotated = [[0] * n for _ in range(m)]

        for i in range(n):
            for j in range(m):
                rotated[m - j - 1][i] = grid[i][j]

        return rotated

    def solve(self, grid):
        n = len(grid)
        m = len(grid[0]) if n > 0 else 0
        res = n * m

        for i in range(n - 1):
            for j in range(m - 1):
                res = min(
                    res,
                    self.minimumArea(grid, 0, i, 0, m - 1)
                    + self.minimumArea(grid, i + 1, n - 1, 0, j)
                    + self.minimumArea(grid, i + 1, n - 1, j + 1, m - 1),
                )

                res = min(
                    res,
                    self.minimumArea(grid, 0, i, 0, j)
                    + self.minimumArea(grid, 0, i, j + 1, m - 1)
                    + self.minimumArea(grid, i + 1, n - 1, 0, m - 1),
                )

        for i in range(n - 2):
            for j in range(i + 1, n - 1):
                res = min(
                    res,
                    self.minimumArea(grid, 0, i, 0, m - 1)
                    + self.minimumArea(grid, i + 1, j, 0, m - 1)
                    + self.minimumArea(grid, j + 1, n - 1, 0, m - 1),
                )

        return res

    def minimumSum(self, grid: List[List[int]]) -> int:
        rgrid = self.rotate(grid)
        return min(self.solve(grid), self.solve(rgrid))